New nex-go

jonbarrow opened the pull request:
Draft since lots of things are changing still, but wanted to get this into more peoples hands

The goal of this PR is to bring this library up to date with the new `nex-go` rewrite in the works https://github.com/PretendoNetwork/nex-go/pull/42. Eventually I would also like to refactor all the methods implemented here which assume a PID is a uint32. A PID is not always a uint32, it can be a uint64 on modern clients (the Switch)

Currently this library does still assume a client is a legacy client when reading PIDs, but we now have the ability to support modern clients here

jonbarrow committed:
work with new nex-go

jonbarrow committed:
use new PID type

jonbarrow committed:
use new PID, DateTime and method returns

jonbarrow committed:
forgot to save a file

jonbarrow committed:
updated to new stream functions

jonbarrow committed:
updated StationURL handling

jonbarrow committed:
export most helpers rather than use setters

jonbarrow committed:
export SecureStationURL in TicketGranting

DaniElectra committed:
Add ContainsPID helper

slices.Cotains doesn't work when checking a list of PIDs, since the PIDs
will usually point to different locations even if the underlying value
is the same.

DaniElectra committed:
Use ServerInterface and ClientInterface

There are some places where the PRUDP implementation is still needed,
but they aren't needed for HPP.

DaniElectra committed:
Update README.md

DaniElectra committed:
Update with new virtual ports handling

DaniElectra committed:
Update with PasswordFromPID on ServerInterface

jonbarrow committed:
merge: fix merge conflict

jonbarrow committed:
ticket-granting: update to new PasswordFromPID handler

jonbarrow committed:
most common protocols now use interfaces

jonbarrow commented:
Most common implementations now use the new protocol interfaces. The exceptions to this are Secure Connection and Matchmaking

Secure Connection *does* technically use the interface, but there are places where it just always assumes a `PRUDPServer` and casts, which makes it not compatible with the `HPPServer`

Matchmaking does not use the interface at all, which I would like to fix at some point as it's now the only common implementation to *not* do it this way

It might get verbose, but we can make use of Go's type assertions to work around these casts. For example in Matchmaking

```go
server.OnClientRemoved(func(client *nex.PRUDPClient) {
	fmt.Println("Leaving")
	common_globals.RemoveClientFromAllSessions(client)
})
```

Could become

```go
if server, ok := server.(*nex.PRUDPServer); ok {
	server.OnClientRemoved(func(client *nex.PRUDPClient) {
		fmt.Println("Leaving")
		common_globals.RemoveClientFromAllSessions(client)
	})
}
```

So that the `OnClientRemoved` call only happens if `server` asserts to a `PRUDPServer`. We would need *many* of these assertions though, for both the server *and* the client interfaces

DaniElectra committed:
matchmaking: Use Interface for common protocol creation

DaniElectra commented:
I think we should be fine right now by assigning the interfaces since HPP isn't expected to work with this. Once we implement websockets, we can remove the casts and use a common interface for PRUDP and websockets for these cases

DaniElectra commented:
btw I have noticed that there is some inconsistent naming for the matchmaking protocols. For example, on nex-protocols-go we use `match-making` and `match-making-ext` but on the common protocols they are named `matchmaking` and `matchmaking-ext`. I think we should rename one of them for better consistency

jonbarrow committed:
ticket-granting: update RVConnectionData

jonbarrow committed:
Merge branch 'nex-go-rewrite' of https://github.com/PretendoNetwork/nex-protocols-common-go into nex-go-rewrite

jonbarrow commented:
> I think we should be fine right now by assigning the interfaces since HPP isn't expected to work with this

I would like at some point to make HPP compatible, since it's my understanding that HPP is supposed to be a drop-in replacement for NEX, just over HTTP. Allowing for access to all protocols/methods normally accessible in the NEX server. But I agree that it's fine for right now. Our use case does not require this, it's mostly a "something nice we should do in the future" thing

> btw I have noticed that there is some inconsistent naming for the matchmaking protocols. For example, on nex-protocols-go we use `match-making` and `match-making-ext` but on the common protocols they are named `matchmaking` and `matchmaking-ext`. I think we should rename one of them for better consistency

I agree. I had not noticed this. I'll update common since the protocols lib is correct

jonbarrow committed:
renamed matchmaking and matchmaking-ext to match protocol lib

jonbarrow committed:
match-making: spacing issue in protocol.go

jonbarrow committed:
ticket-granting: fix RVConnectionData

jonbarrow committed:
ticket-granting: fix ineffectual assignment to userPID

DaniElectra committed:
ticket-granting: Clear StationURLSpecialProtocols

This field doesn't have a station URL set usually.

DaniElectra committed:
Update PRUDP cast comments

Now that the websockets server is integrated into PRUDPServer, we have
to question if we even want to remove these casts, since it doesn't make
much sense for HPP.

Also updaete SecureConnection::Register to support TCP addresses on the
client for websocket implementations.

DaniElectra committed:
README: Update according to latest changes

DaniElectra commented:
I'm thinking if we should rename the common protocol creation methods too, like for the raw protocols where we changed them to `NewProtocol`. This is an example for creating a common secure protocol:

```go
secureProtocol := secure.NewProtocol(globals.SecureServer)
common_secure.NewCommonSecureConnectionProtocol(secureProtocol)
```

Maybe we could rename them to something like `NewCommonProtocol` to simplify?

```go
secureProtocol := secure.NewProtocol(globals.SecureServer)
common_secure.NewCommonProtocol(secureProtocol)
```

jonbarrow commented:
Yes that looks a lot better imo

DaniElectra committed:
Rename common protocols to CommonProtocol

This reduces the redundancy and follows better the Go style.

DaniElectra committed:
README: Don't update all commonTicketGrantingProtocol fields

DaniElectra committed:
Update RMC method creation

jonbarrow committed:
chore: update to new types

jonbarrow commented:
@DaniElectra I updated everything to support the new type system. It's basically untested though, just did enough to get Go to stop showing errors/warnings

HPP support also needs to be re-added. Right now it assumes a `PRUDPConnection`. In most cases though this is just to remove a type assertion for the `server` variable, because there's TONS of places where we just do blind assertions without checking (which WILL lead to panics if the wrong client interacts with some protocols)

A solution to the `PRUDPConnection` assumption would be to just check the type of `packet.Sender()`. If it's a `PRUDPConnection` then use the connections endpoint to get the server like it is now, otherwise if it's a `HPPClient` use the server on the client

jonbarrow committed:
ticket-granting: update to use new account methods

jonbarrow committed:
chore: nex.Errors -> nex.ResultCodes

DaniElectra commented on ticket-granting/protocol.go:
```diff
@@ -2,51 +2,54 @@ package ticket_granting
+	BuildName                  *types.String
+	allowInsecureLoginMethod   bool
+	SessionKeyLength           int
+	AccountDetailsByPID        func(pid *types.PID) (*nex.Account, uint32)
+	AccountDetailsByUsername   func(username string) (*nex.Account, uint32)
+	SecureServerAccount        *nex.Account
```
Maybe we could take these values from the PRUDP endpoint, once HPP checks are sorted out?

jonbarrow replied:
I'm open to the idea if you can propose an API. What do you propose for getting it from the endpoint? This protocol is supposed to be used on the authentication endpoint, which uses the authentication account. If we pull that endpoints account it won't be the correct one, since it needs the secure server account

DaniElectra replied:
You are right about the secure server account, we can keep that. but maybe we could at least use the `AccountDetails` functions from the endpoint?

```go
	sourceAccount, errorCode := endpoint.AccountDetailsByUsername(strUserName.Value)
	if errorCode != nil && errorCode.ResultCode != nex.ResultCodes.RendezVous.InvalidUsername {
		// * Some other error happened
		return nil, errorCode.ResultCode
	}

	targetAccount, errorCode := endpoint.AccountDetailsByUsername(commonProtocol.SecureServerAccount.Username)
	if errorCode != nil && errorCode.ResultCode != nex.ResultCodes.RendezVous.InvalidUsername {
		// * Some other error happened
		return nil, errorCode.ResultCode
	}
```

jonbarrow replied:
Sounds good to me :+1:

DaniElectra committed:
Fix typos and use account getters from endpoint

DaniElectra committed:
refactor: Update ServerInterface to EndpointInterface

Also update method handlers to use the new nex.Error return.

DaniElectra committed:
refactor: Remove common protocols global variables

jonbarrow committed:
nat-traversal: fixed comment

jonbarrow committed:
nat-traversal: fixed ReportNATTraversalResultDetail method ID

jonbarrow committed:
update: added OnAfter event handlers for common methods

jonbarrow committed:
matchmake-extension: rename postpone method handlers

DaniElectra committed:
match-making: Fix wrong session ownership check

DaniElectra committed:
matchmake-extension: Implement JoinMatchmakeSessionEx

And some bugfixes.

PabloMK7 committed:
Removing lingering "Leaving" print

PabloMK7 committed:
Remove unneeded import

DaniElectra committed:
match-making: Remove "Leaving" print

In case something fails here, this can be logged on the server directly for debugging

DaniElectra committed:
match-making: Decrement ParticipationCount when leaving a session

DaniElectra committed:
chore: Use nex-go constants

DaniElectra committed:
match-making: Use MutexSlice for connection IDs

DaniElectra committed:
secure-connection: Fix public station flag check

DaniElectra committed:
match-making: Fix race condition on FindOtherConnectionID

DaniElectra committed:
match-making: Change the host if the owner is leaving

DaniElectra committed:
match-making: Update ParticipationCount with ConnectionIDs size

DaniElectra committed:
match-making: Set original owner on ownership change notification

DaniElectra committed:
chore: Update module version to v2
